File: python/tests/test_voice_text.py
"""test_voice_text.py

Asserts the Python SentenceSplitter / LanguageSpanSplitter / GeezRomanizer /
ToneShaper / NchltPhonemizer ports against the same golden files the C#
reference generates.

Every case in these fixtures is adversarial. The splitter fixture carries a
decimal point and a domain name that must NOT split next to a danda and a CJK
stop that must; the Ge'ez fixture carries the numerals that used to romanise as
syllables; the tone fixture separates the biquad (bit-reproducible) from the
coefficient derivation (pow/sin/cos, which no language guarantees to the last
bit).
"""
from __future__ import annotations

import json
import pathlib
import struct

import pytest

from circle_ai.voice_text import (
    MAX_CHARS_PER_SEGMENT,
    WARM,
    NchltPhonemizer,
    apply_tone_shaper,
    biquad,
    is_ethiopic,
    is_foreign_word,
    low_shelf_coefficients,
    peaking_coefficients,
    romanize,
    split_language_spans,
    split_sentences,
    to_spoken_form,
)

# tests/ -> python/ -> CircleAI/ -> fixtures/
FIXTURES = pathlib.Path(__file__).resolve().parents[2] / "fixtures"


def _read(name: str) -> dict:
    return json.loads((FIXTURES / name).read_text(encoding="utf-8"))


def _fround(v: float) -> float:
    return struct.unpack("<f", struct.pack("<f", v))[0]


@pytest.fixture(scope="module")
def splitter_fixture() -> dict:
    return _read("voice_sentence_splitter.json")


@pytest.fixture(scope="module")
def spans_fixture() -> dict:
    return _read("voice_language_spans.json")


@pytest.fixture(scope="module")
def geez_fixture() -> dict:
    return _read("voice_geez_romanizer.json")


@pytest.fixture(scope="module")
def tone_fixture() -> dict:
    return _read("voice_tone_shaper.json")


@pytest.fixture(scope="module")
def nchlt_fixture() -> dict:
    return _read("voice_nchlt_phonemizer.json")


# ── SentenceSplitter ────────────────────────────────────────────────────────


def test_sentence_splitter_matches_reference(splitter_fixture: dict) -> None:
    assert MAX_CHARS_PER_SEGMENT == splitter_fixture["maxCharsPerSegment"]
    for c in splitter_fixture["cases"]:
        got = [
            {"text": s.text, "trailingPauseMs": s.trailing_pause_ms}
            for s in split_sentences(c["text"])
        ]
        assert got == c["segments"], f"segments for {c['name']}"


def test_splits_scripts_that_do_not_punctuate_in_latin(splitter_fixture: dict) -> None:
    # A Latin-only terminator list under-splits for about a billion people and
    # fails silently — the paragraph simply runs together.
    by_name = {c["name"]: c for c in splitter_fixture["cases"]}
    for name in ("devanagari-danda", "urdu-full-stop", "cjk-no-space", "khmer-khan"):
        assert len(split_sentences(by_name[name]["text"])) > 1, f"{name} must split"


def test_does_not_split_a_decimal_or_a_domain(splitter_fixture: dict) -> None:
    by_name = {c["name"]: c for c in splitter_fixture["cases"]}
    for name in ("decimal-point", "domain-name"):
        assert len(split_sentences(by_name[name]["text"])) == 2, name


def test_last_segment_has_no_trailing_pause(splitter_fixture: dict) -> None:
    for c in splitter_fixture["cases"]:
        got = split_sentences(c["text"])
        if got:
            assert got[-1].trailing_pause_ms == 0, c["name"]


# ── LanguageSpanSplitter ────────────────────────────────────────────────────


def test_language_spans_match_reference(spans_fixture: dict) -> None:
    for c in spans_fixture["split"]:
        got = [
            {"text": s.text, "isForeign": s.is_foreign}
            for s in split_language_spans(c["text"])
        ]
        assert got == c["spans"], f"spans for {c['text']}"


def test_to_spoken_form_matches_reference(spans_fixture: dict) -> None:
    for c in spans_fixture["toSpokenForm"]:
        assert to_spoken_form(c["input"]) == c["output"], f"spoken form of {c['input']}"


def test_is_foreign_word_matches_reference(spans_fixture: dict) -> None:
    for c in spans_fixture["isForeignWord"]:
        assert is_foreign_word(c["word"]) == c["foreign"], c["word"]
    # The conservatism is the contract, not an accident: an ordinary lowercase
    # English word must NOT be flagged, because guessing wrong mispronounces a
    # native word to fix a foreign one.
    assert is_foreign_word("hello") is False
    assert is_foreign_word("Ngiyabonga") is False


# ── GeezRomanizer ───────────────────────────────────────────────────────────


def test_is_ethiopic_matches_reference(geez_fixture: dict) -> None:
    for c in geez_fixture["isEthiopic"]:
        assert is_ethiopic(c["text"]) == c["ethiopic"], repr(c["text"])


def test_romanize_matches_reference(geez_fixture: dict) -> None:
    for c in geez_fixture["romanize"]:
        assert romanize(c["input"]) == c["output"], repr(c["input"])


def test_numerals_are_dropped_not_spoken() -> None:
    # The eight-per-consonant layout stops at U+1357. Sizing the range check off
    # the consonant table swept seven numerals back into the syllabary, and they
    # came out as sound, so nothing failed.
    assert romanize("፩፪፫") == ""
    assert romanize("ፘፙፚ") == "ryamyafya", "the three LONE syllables are not a row"


# ── ToneShaper ──────────────────────────────────────────────────────────────


def _assert_close(got: float, want: float, tol: float, what: str) -> None:
    scale = max(1.0, abs(want))
    assert abs(got - want) <= tol * scale, f"{what}: got {got}, want {want}"


def test_tone_shaper_settings(tone_fixture: dict) -> None:
    # Field by field, and NOT against the whole fixture object: the shelf slope
    # is a private constant of the filter, not a setting anyone may pass in.
    s = tone_fixture["settings"]
    assert WARM.low_shelf_hz == s["lowShelfHz"]
    assert WARM.low_shelf_db == s["lowShelfDb"]
    assert WARM.presence_hz == s["presenceHz"]
    assert WARM.presence_db == s["presenceDb"]
    assert WARM.presence_q == s["presenceQ"]
    assert s["lowShelfSlope"] == 0.9


def test_tone_shaper_coefficients(tone_fixture: dict) -> None:
    # 1e-9 relative, not exact: pow, sin and cos are not bit-identical across
    # languages, and pretending otherwise makes a flaky test rather than a strict
    # one.
    tol = tone_fixture["coefficientTolerance"]
    for c in tone_fixture["coefficients"]:
        got = {
            "lowShelf": low_shelf_coefficients(WARM, c["sampleRate"]),
            "peaking": peaking_coefficients(WARM, c["sampleRate"]),
        }
        for name in ("lowShelf", "peaking"):
            b, a = got[name]
            for i in range(3):
                _assert_close(b[i], c[name]["b"][i], tol, f"{name} b[{i}] @{c['sampleRate']}")
                _assert_close(a[i], c[name]["a"][i], tol, f"{name} a[{i}] @{c['sampleRate']}")


def test_tone_shaper_waveform(tone_fixture: dict) -> None:
    # The biquad is add and multiply on doubles, so THIS half is expected to
    # agree everywhere. Driving it from the fixture's own coefficients keeps the
    # transcendental functions out of the comparison.
    w = tone_fixture["waveform"]
    coeffs = next(c for c in tone_fixture["coefficients"] if c["sampleRate"] == w["sampleRate"])

    x = [_fround(v) for v in w["input"]]
    before = max(abs(v) for v in x)
    biquad(x, coeffs["lowShelf"]["b"], coeffs["lowShelf"]["a"])
    biquad(x, coeffs["peaking"]["b"], coeffs["peaking"]["a"])
    after = max(abs(v) for v in x)
    if after > 0 and after > before:
        g = _fround(before / after)
        x = [_fround(v * g) for v in x]

    for i, want in enumerate(w["output"]):
        _assert_close(x[i], want, tone_fixture["waveformTolerance"], f"sample {i}")


def test_silence_stays_silent(tone_fixture: dict) -> None:
    silence = [0.0] * len(tone_fixture["silenceStaysSilent"])
    apply_tone_shaper(silence, tone_fixture["waveform"]["sampleRate"])
    assert silence == tone_fixture["silenceStaysSilent"]


def test_both_filters_are_applied(tone_fixture: dict) -> None:
    # A port that dropped the presence dip would still change the waveform, so
    # "it moved" proves nothing — the two stages must differ from each other.
    rate = tone_fixture["waveform"]["sampleRate"]
    x = [_fround(v) for v in tone_fixture["waveform"]["input"]]
    only_shelf = list(x)
    apply_tone_shaper(x, rate)
    b, a = low_shelf_coefficients(WARM, rate)
    biquad(only_shelf, b, a)
    assert any(abs(x[i] - only_shelf[i]) > 1e-4 for i in range(len(x))), (
        "the presence dip made no difference — it was not applied"
    )


# ── NchltPhonemizer ─────────────────────────────────────────────────────────


def _make(fixture: dict) -> NchltPhonemizer:
    return NchltPhonemizer.from_text(
        fixture["dict"], fixture["rules"], fixture["phoneMap"],
        fixture["graphMap"], fixture["gnulls"],
    )


def test_nchlt_matches_reference(nchlt_fixture: dict) -> None:
    for c in nchlt_fixture["cases"]:
        p = _make(nchlt_fixture)
        assert p.phonemize(c["text"]) == c["phones"], f"phones for {c['name']}"
        assert p.last_rule_predicted_words == c["rulePredictedWords"], c["name"]
        assert p.last_unknown_graphemes == c["unknownGraphemes"], c["name"]


def test_nchlt_predict_word(nchlt_fixture: dict) -> None:
    for c in nchlt_fixture["predictWord"]:
        assert _make(nchlt_fixture).predict_word(c["word"]) == c["phones"], c["word"]


def test_dictionary_beats_the_rules(nchlt_fixture: dict) -> None:
    # Both paths can pronounce this word. The dictionary must win, and the rule
    # counter must show it did — the counter is the only evidence of which path
    # ran, and a port that always predicted would still return sensible phones.
    p = _make(nchlt_fixture)
    p.phonemize("sawubona")
    assert p.last_rule_predicted_words == 0


def test_unknown_grapheme_is_reported(nchlt_fixture: dict) -> None:
    p = _make(nchlt_fixture)
    p.phonemize("azb")
    assert p.last_unknown_graphemes == ["z"]

File: c/tests/test_voice_text.c
/*
 * test_voice_text.c — asserts the C SentenceSplitter / LanguageSpanSplitter /
 * GeezRomanizer / ToneShaper / NchltPhonemizer ports against the same answers
 * the C# reference generates.
 *
 * THE EXPECTED VALUES LIVE IN voice_text_expected.h AS LITERALS, not parsed from
 * the fixtures. This port has no JSON reader and will not vendor one; the other
 * eight ports read the fixtures directly, which is stronger. That makes this the
 * one port where a fixture can drift without a test failing.
 *
 * The header is GENERATED rather than hand-typed, because 300 lines of
 * transcription is 300 chances to mistype a codepoint. Regenerate it after any
 * fixture change, IN THE SAME COMMIT:
 *
 *   python tools/gen_c_voice_expected.py fixtures c/tests/voice_text_expected.h
 *
 * The cases are adversarial: a decimal point and a domain name that must NOT
 * split next to a danda and a CJK stop that must; the Ethiopic numerals that
 * used to romanise as syllables; and a tone fixture that separates the biquad
 * (bit-reproducible) from the coefficient derivation (pow/sin/cos, which no
 * language guarantees to the last bit).
 */

#include "circle_ai/voice_text.h"

#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static int failures = 0;
static int checks = 0;

static void check(int cond, const char *what)
{
    checks++;
    if (!cond) { printf("  FAIL: %s\n", what); failures++; }
}

static void check_str(const char *got, const char *want, const char *what)
{
    checks++;
    if (!got || strcmp(got, want) != 0) {
        printf("  FAIL: %s — got \"%s\", want \"%s\"\n", what, got ? got : "(null)", want);
        failures++;
    }
}

static void check_close(double got, double want, double tol, const char *what)
{
    checks++;
    double scale = fabs(want) > 1.0 ? fabs(want) : 1.0;
    if (fabs(got - want) > tol * scale) {
        printf("  FAIL: %s — got %.17g, want %.17g\n", what, got, want);
        failures++;
    }
}

/* Shapes the generated header fills in. */
typedef struct { const char *text; int pause; } seg_expect;
typedef struct {
    const char *name; const char *text;
    const seg_expect *segments; size_t count;
} splitter_case;

typedef struct { const char *text; int is_foreign; } span_expect;
typedef struct { const char *text; const span_expect *spans; size_t count; } spans_case;

typedef struct { const char *input; const char *output; } pair_case;
typedef struct { const char *input; int flag; } flag_case;

typedef struct {
    double low_shelf_hz, low_shelf_db, presence_hz, presence_db, presence_q, slope;
} settings_expect;
typedef struct {
    int rate;
    double ls_b[3], ls_a[3], pk_b[3], pk_a[3];
} coeff_case;

typedef struct {
    const char *name; const char *text;
    const char *const *phones; size_t phone_count;
    int rule_predicted; const char *const *unknown; size_t unknown_count;
} nchlt_case;
typedef struct {
    const char *word; const char *const *phones; size_t phone_count;
} predict_case;

#include "voice_text_expected.h"

#define COUNT(a) (sizeof (a) / sizeof (a)[0])

/* ---- SentenceSplitter --------------------------------------------------- */

static void test_sentence_splitter(void)
{
    printf("SentenceSplitter\n");

    check(CA_MAX_CHARS_PER_SEGMENT == SPLITTER_MAX_CHARS, "MaxCharsPerSegment");

    for (size_t c = 0; c < COUNT(SPLITTER_CASES); c++) {
        ca_speech_segment_t got[32];
        size_t n = ca_split_sentences(SPLITTER_CASES[c].text, got, 32);

        char what[160];
        snprintf(what, sizeof what, "%s segment count", SPLITTER_CASES[c].name);
        checks++;
        if (n != SPLITTER_CASES[c].count) {
            printf("  FAIL: %s — got %zu segments, want %zu\n",
                   what, n, SPLITTER_CASES[c].count);
            failures++;
            ca_speech_segments_free(got, n < 32 ? n : 32);
            continue;
        }

        for (size_t i = 0; i < n; i++) {
            snprintf(what, sizeof what, "%s segment %zu text", SPLITTER_CASES[c].name, i);
            check_str(got[i].text, SPLITTER_CASES[c].segments[i].text, what);
            snprintf(what, sizeof what, "%s segment %zu pause", SPLITTER_CASES[c].name, i);
            check(got[i].trailing_pause_ms == SPLITTER_CASES[c].segments[i].pause, what);
        }
        ca_speech_segments_free(got, n);
    }
}

static const splitter_case *find_case(const char *name)
{
    for (size_t i = 0; i < COUNT(SPLITTER_CASES); i++)
        if (strcmp(SPLITTER_CASES[i].name, name) == 0) return &SPLITTER_CASES[i];
    return NULL;
}

static void test_splits_non_latin_scripts(void)
{
    printf("Scripts that do not punctuate in Latin still split\n");

    /* A Latin-only terminator list under-splits for about a billion people and
     * fails silently — the paragraph simply runs together. */
    static const char *const names[] = {
        "devanagari-danda", "urdu-full-stop", "cjk-no-space", "khmer-khan",
    };

    for (size_t i = 0; i < COUNT(names); i++) {
        const splitter_case *c = find_case(names[i]);
        checks++;
        if (!c) { printf("  FAIL: no case named %s\n", names[i]); failures++; continue; }

        ca_speech_segment_t got[32];
        size_t n = ca_split_sentences(c->text, got, 32);
        if (n < 2) { printf("  FAIL: %s produced %zu segments\n", names[i], n); failures++; }
        ca_speech_segments_free(got, n < 32 ? n : 32);
    }
}

static void test_does_not_split_decimal_or_domain(void)
{
    printf("A decimal point and a domain name do not split\n");

    static const char *const names[] = { "decimal-point", "domain-name" };
    for (size_t i = 0; i < COUNT(names); i++) {
        const splitter_case *c = find_case(names[i]);
        checks++;
        if (!c) { printf("  FAIL: no case named %s\n", names[i]); failures++; continue; }

        ca_speech_segment_t got[32];
        size_t n = ca_split_sentences(c->text, got, 32);
        if (n != 2) { printf("  FAIL: %s produced %zu segments, want 2\n", names[i], n); failures++; }
        ca_speech_segments_free(got, n < 32 ? n : 32);
    }
}

/* ---- LanguageSpanSplitter ----------------------------------------------- */

static void test_language_spans(void)
{
    printf("LanguageSpanSplitter\n");

    for (size_t c = 0; c < COUNT(SPANS_CASES); c++) {
        ca_language_span_t got[16];
        size_t n = ca_split_language_spans(SPANS_CASES[c].text, got, 16);

        char what[160];
        snprintf(what, sizeof what, "span count for %s", SPANS_CASES[c].text);
        checks++;
        if (n != SPANS_CASES[c].count) {
            printf("  FAIL: %s — got %zu, want %zu\n", what, n, SPANS_CASES[c].count);
            failures++;
            ca_language_spans_free(got, n < 16 ? n : 16);
            continue;
        }

        for (size_t i = 0; i < n; i++) {
            snprintf(what, sizeof what, "span %zu text of %s", i, SPANS_CASES[c].text);
            check_str(got[i].text, SPANS_CASES[c].spans[i].text, what);
            snprintf(what, sizeof what, "span %zu flag of %s", i, SPANS_CASES[c].text);
            check(got[i].is_foreign == SPANS_CASES[c].spans[i].is_foreign, what);
        }
        ca_language_spans_free(got, n);
    }

    for (size_t i = 0; i < COUNT(SPOKEN_CASES); i++) {
        char *got = ca_to_spoken_form(SPOKEN_CASES[i].input);
        char what[160];
        snprintf(what, sizeof what, "spoken form of %s", SPOKEN_CASES[i].input);
        check_str(got, SPOKEN_CASES[i].output, what);
        free(got);
    }

    for (size_t i = 0; i < COUNT(FOREIGN_CASES); i++) {
        char what[160];
        snprintf(what, sizeof what, "isForeignWord(%s)", FOREIGN_CASES[i].input);
        check(ca_is_foreign_word(FOREIGN_CASES[i].input) == FOREIGN_CASES[i].flag, what);
    }

    /* The conservatism is the contract, not an accident: guessing wrong
     * mispronounces a native word to fix a foreign one. */
    check(!ca_is_foreign_word("hello"), "an ordinary word is not foreign");
    check(!ca_is_foreign_word("Ngiyabonga"), "a capitalised native word is not foreign");
}

/* ---- GeezRomanizer ------------------------------------------------------ */

static void test_geez(void)
{
    printf("GeezRomanizer\n");

    for (size_t i = 0; i < COUNT(ETHIOPIC_CASES); i++) {
        char what[160];
        snprintf(what, sizeof what, "isEthiopic(%s)", ETHIOPIC_CASES[i].input);
        check(ca_is_ethiopic(ETHIOPIC_CASES[i].input) == ETHIOPIC_CASES[i].flag, what);
    }

    for (size_t i = 0; i < COUNT(ROMANIZE_CASES); i++) {
        char *got = ca_geez_romanize(ROMANIZE_CASES[i].input);
        char what[160];
        snprintf(what, sizeof what, "romanize(%s)", ROMANIZE_CASES[i].input);
        check_str(got, ROMANIZE_CASES[i].output, what);
        free(got);
    }

    /* The eight-per-consonant layout stops at U+1357. Sizing the range check off
     * the consonant table swept seven numerals back into the syllabary, and they
     * came out as sound, so nothing failed. */
    char *numerals = ca_geez_romanize("\xe1\x8d\xa9\xe1\x8d\xaa\xe1\x8d\xab");
    check_str(numerals, "", "the numerals have no sound to render");
    free(numerals);

    char *lone = ca_geez_romanize("\xe1\x8d\x98\xe1\x8d\x99\xe1\x8d\x9a");
    check_str(lone, "ryamyafya", "the three LONE syllables are not a row of eight");
    free(lone);
}

/* ---- ToneShaper --------------------------------------------------------- */

static float peak_of(const float *x, size_t n)
{
    float p = 0;
    for (size_t i = 0; i < n; i++) { float a = x[i] < 0 ? -x[i] : x[i]; if (a > p) p = a; }
    return p;
}

static void test_tone_shaper(void)
{
    printf("ToneShaper\n");

    ca_tone_shaper_t warm = ca_tone_shaper_warm();
    check(warm.low_shelf_hz == TONE_SETTINGS.low_shelf_hz, "lowShelfHz");
    check(warm.low_shelf_db == TONE_SETTINGS.low_shelf_db, "lowShelfDb");
    check(warm.presence_hz == TONE_SETTINGS.presence_hz, "presenceHz");
    check(warm.presence_db == TONE_SETTINGS.presence_db, "presenceDb");
    check(warm.presence_q == TONE_SETTINGS.presence_q, "presenceQ");
    check(TONE_SETTINGS.slope == 0.9, "the shelf slope is fixed at 0.9");

    /* 1e-9 relative, not exact: pow, sin and cos are not bit-identical across
     * languages, and pretending otherwise makes a flaky test, not a strict one. */
    for (size_t i = 0; i < COUNT(COEFF_CASES); i++) {
        ca_biquad_coefficients_t ls = ca_low_shelf_coefficients(&warm, COEFF_CASES[i].rate);
        ca_biquad_coefficients_t pk = ca_peaking_coefficients(&warm, COEFF_CASES[i].rate);
        for (int k = 0; k < 3; k++) {
            char what[96];
            snprintf(what, sizeof what, "lowShelf b[%d] at %d", k, COEFF_CASES[i].rate);
            check_close(ls.b[k], COEFF_CASES[i].ls_b[k], TONE_COEFFICIENT_TOLERANCE, what);
            snprintf(what, sizeof what, "lowShelf a[%d] at %d", k, COEFF_CASES[i].rate);
            check_close(ls.a[k], COEFF_CASES[i].ls_a[k], TONE_COEFFICIENT_TOLERANCE, what);
            snprintf(what, sizeof what, "peaking b[%d] at %d", k, COEFF_CASES[i].rate);
            check_close(pk.b[k], COEFF_CASES[i].pk_b[k], TONE_COEFFICIENT_TOLERANCE, what);
            snprintf(what, sizeof what, "peaking a[%d] at %d", k, COEFF_CASES[i].rate);
            check_close(pk.a[k], COEFF_CASES[i].pk_a[k], TONE_COEFFICIENT_TOLERANCE, what);
        }
    }

    /* The biquad is add and multiply on doubles, so THIS half is expected to
     * agree everywhere. Driving it from the fixture's own coefficients keeps the
     * transcendental functions out of the comparison. */
    const coeff_case *entry = NULL;
    for (size_t i = 0; i < COUNT(COEFF_CASES); i++)
        if (COEFF_CASES[i].rate == TONE_SAMPLE_RATE) { entry = &COEFF_CASES[i]; break; }
    check(entry != NULL, "coefficients for the waveform's sample rate");
    if (!entry) return;

    size_t n = COUNT(TONE_INPUT);
    float *x = (float *)malloc(n * sizeof(float));
    if (!x) { check(0, "allocation"); return; }
    for (size_t i = 0; i < n; i++) x[i] = (float)TONE_INPUT[i];

    ca_biquad_coefficients_t ls, pk;
    memcpy(ls.b, entry->ls_b, sizeof ls.b);
    memcpy(ls.a, entry->ls_a, sizeof ls.a);
    memcpy(pk.b, entry->pk_b, sizeof pk.b);
    memcpy(pk.a, entry->pk_a, sizeof pk.a);

    float before = peak_of(x, n);
    ca_biquad(x, n, &ls);
    ca_biquad(x, n, &pk);
    float after = peak_of(x, n);
    if (after > 0.0f && after > before) {
        float g = before / after;
        for (size_t i = 0; i < n; i++) x[i] *= g;
    }

    for (size_t i = 0; i < COUNT(TONE_OUTPUT); i++) {
        char what[64];
        snprintf(what, sizeof what, "sample %zu", i);
        check_close(x[i], TONE_OUTPUT[i], TONE_WAVEFORM_TOLERANCE, what);
    }

    /* A port that dropped the presence dip would still change the waveform, so
     * "it moved" proves nothing — the two stages must differ from each other. */
    float *both = (float *)malloc(n * sizeof(float));
    float *only_shelf = (float *)malloc(n * sizeof(float));
    if (both && only_shelf) {
        for (size_t i = 0; i < n; i++) both[i] = only_shelf[i] = (float)TONE_INPUT[i];
        ca_apply_tone_shaper(both, n, TONE_SAMPLE_RATE, &warm);
        ca_biquad_coefficients_t derived = ca_low_shelf_coefficients(&warm, TONE_SAMPLE_RATE);
        ca_biquad(only_shelf, n, &derived);

        int differs = 0;
        for (size_t i = 0; i < n; i++)
            if (fabs((double)both[i] - only_shelf[i]) > 1e-4) { differs = 1; break; }
        check(differs, "the presence dip was applied, not just the shelf");
    }
    free(both);
    free(only_shelf);
    free(x);

    /* A silent buffer must come back untouched: Apply bails when the peak is 0,
     * and a port that divided by that peak would produce NaN. */
    float silence[TONE_SILENCE_COUNT] = { 0 };
    ca_apply_tone_shaper(silence, TONE_SILENCE_COUNT, TONE_SAMPLE_RATE, &warm);
    int quiet = 1;
    for (size_t i = 0; i < TONE_SILENCE_COUNT; i++) if (silence[i] != 0.0f) quiet = 0;
    check(quiet, "silence stays silent rather than dividing by its peak");
}

/* ---- NchltPhonemizer ---------------------------------------------------- */

static ca_nchlt_phonemizer *make_phonemizer(void)
{
    return ca_nchlt_new(NCHLT_DICT, NCHLT_RULES, NCHLT_PHONE_MAP,
                            NCHLT_GRAPH_MAP, NCHLT_GNULLS);
}

static void test_nchlt(void)
{
    printf("NchltPhonemizer\n");

    for (size_t c = 0; c < COUNT(NCHLT_CASES); c++) {
        ca_nchlt_phonemizer *p = make_phonemizer();
        check(p != NULL, "phonemizer loads");
        if (!p) return;

        const char *got[32];
        size_t n = ca_nchlt_phonemize(p, NCHLT_CASES[c].text, got, 32);

        char what[160];
        snprintf(what, sizeof what, "%s phone count", NCHLT_CASES[c].name);
        checks++;
        if (n != NCHLT_CASES[c].phone_count) {
            printf("  FAIL: %s — got %zu phones, want %zu\n",
                   what, n, NCHLT_CASES[c].phone_count);
            failures++;
        } else {
            for (size_t i = 0; i < n; i++) {
                snprintf(what, sizeof what, "%s phone %zu", NCHLT_CASES[c].name, i);
                check_str(got[i], NCHLT_CASES[c].phones[i], what);
            }
        }

        snprintf(what, sizeof what, "%s rulePredictedWords", NCHLT_CASES[c].name);
        check((int)ca_nchlt_last_rule_predicted_words(p) == NCHLT_CASES[c].rule_predicted,
              what);

        const char *const *unknown = NULL;
        size_t un = ca_nchlt_last_unknown_graphemes(p, &unknown);
        snprintf(what, sizeof what, "%s unknown count", NCHLT_CASES[c].name);
        checks++;
        if (un != NCHLT_CASES[c].unknown_count) {
            printf("  FAIL: %s — got %zu unknown, want %zu\n",
                   what, un, NCHLT_CASES[c].unknown_count);
            failures++;
        } else {
            for (size_t i = 0; i < un; i++) {
                snprintf(what, sizeof what, "%s unknown %zu", NCHLT_CASES[c].name, i);
                check_str(unknown[i], NCHLT_CASES[c].unknown[i], what);
            }
        }

        ca_nchlt_free(p);
    }

    for (size_t c = 0; c < COUNT(PREDICT_CASES); c++) {
        ca_nchlt_phonemizer *p = make_phonemizer();
        if (!p) return;

        const char *got[32];
        size_t n = ca_nchlt_predict_word(p, PREDICT_CASES[c].word, got, 32);

        char what[160];
        snprintf(what, sizeof what, "predictWord(%s) count", PREDICT_CASES[c].word);
        checks++;
        if (n != PREDICT_CASES[c].phone_count) {
            printf("  FAIL: %s — got %zu, want %zu\n", what, n, PREDICT_CASES[c].phone_count);
            failures++;
        } else {
            for (size_t i = 0; i < n; i++) {
                snprintf(what, sizeof what, "predictWord(%s) phone %zu",
                         PREDICT_CASES[c].word, i);
                check_str(got[i], PREDICT_CASES[c].phones[i], what);
            }
        }
        ca_nchlt_free(p);
    }

    /* Both paths can pronounce this word. The dictionary must win, and the rule
     * counter must show it did — the counter is the only evidence of which path
     * ran, and a port that always predicted would still return sensible phones. */
    {
        ca_nchlt_phonemizer *p = make_phonemizer();
        if (p) {
            const char *got[32];
            ca_nchlt_phonemize(p, "sawubona", got, 32);
            check(ca_nchlt_last_rule_predicted_words(p) == 0,
                  "a catalogued word must not be predicted");
            ca_nchlt_free(p);
        }
    }
}

int main(void)
{
    printf("=== voice text-module parity ===\n");
    test_sentence_splitter();
    test_splits_non_latin_scripts();
    test_does_not_split_decimal_or_domain();
    test_language_spans();
    test_geez();
    test_tone_shaper();
    test_nchlt();
    printf("%d checks, %d failures\n", checks, failures);
    return failures == 0 ? 0 : 1;
}
